cs4660/graph/graph.py

```diff
@@ -0,0 +1,380 @@
+    def __repr__(self):
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+        return self.data == other_node.data
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.data)
+
+class Edge(object):
+
+    def __init__(self, source, destination, weight):
+        self.source = source
+        self.destination = destination
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.source, self.destination, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.source, self.destination, self.weight)
+
+    def __eq__(self, other_node):
+        return self.source == other_node.source and self.destination == other_node.destination and self.weight == other_node.weight
+    
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.source, self.destination, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+        a = len(self.adjacency_list)
+     
+
+
+    def adjacent(self, node_1, node_2):
+    	if node_2 not in self.adjacency_list[node_1] :
+    		if node_1 not in self.adjacency_list[node_2]:
+    			return False
+    		
+    	elif node_2  in self.adjacency_list[node_1] :
+    			return True
+    	
+    	
+
+
+
+    def neighbors(self, node):
+    	list1 =[]
+    	if(len(self.adjacency_list[node])==0):
+    		return list1
+    	elif(len(self.adjacency_list[node])!=0):
+    		return list(self.adjacency_list[node]);
+
+
+
+    def add_node(self, node):
+    	if(node not in self.adjacency_list):
+    		self.adjacency_list[node] ={}
+    		return True
+    	elif(node in self.adjacency_list):
+    		return False
+
+
+    def remove_node(self, node):	
+        if(node not in self.adjacency_list):
+        	return False
+        elif(node in self.adjacency_list):
+        	for i in self.adjacency_list:
+        		self.adjacency_list[i] ={}
+        	return True;
+
+
+    def add_edge(self, edge):
+    	source = edge.source;
+    	destination= edge.destination;
+    	if(destination in self.adjacency_list[source]):
+    		return False;
+    	elif(destination not in self.adjacency_list[source]):
+    		self.adjacency_list[source][destination] = edge.weight;
+    		return True;
+
+
+
+       
+
+    def remove_edge(self, edge):
+    	source = edge.source
+    	destination= edge.destination
+
+    	
+    	if(destination not in self.adjacency_list[source]):
+    		return False
+    	elif(destination in self.adjacency_list[source]):
+    		self.adjacency_list[source] = {} 
+    		self.adjacency_list[destination] ={}
```

Comment:
remove edge should only be removing edge one way, (assume graph is directed graph)